app/controllers/api/observation_sessions_controller.rb

```diff
@@ -2,16 +2,24 @@ module Api
   class ObservationSessionsController < BaseController
     def create
       ActiveRecord::Base.transaction do
-        create_observation_sessions
-        # TODO: upsert?
-        observations = observations_params.map do |observation|
+        
+      session_ids_to_instances = {} 
+      processed_observations = []   
+
+      observations_params.each do | observation |
+          session = session_ids_to_instances.fetch(observation['observation_session_id']) do | session_id |
+            session_ids_to_instances[session_id] = ObservationSession.find_or_create_by(id: session_id, user_id: current_user.id)
+          end
+          
           Observation.find_or_create_by(id: observation['id']) do |record|
             record.details = observation
-            record.observation_session_id = observation['observation_session_id']
+            record.observation_session = session
+            record.save
```

Comment:
I believe find or create will execute the block if it's a create, but not if it's a find, but because the block is passed on to create, the value will be set anyway. I can definitely remove the `save`, but I'm concerned the method doesn't behave as expected when an observation with that ID is sent. https://stackoverflow.com/q/5461315/3896861

Does this method ever need to update an existing observation id?